torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -198,6 +202,102 @@ bool is_numpy_scalar(PyObject* obj) {
           PyArray_IsScalar(obj, Floating));
 }
 
+at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  PyObject *cuda_dict = PyObject_GetAttrString(obj, "__cuda_array_interface__");
+  if (cuda_dict == nullptr) {
+    throw TypeError("attribute `__cuda_array_interface__` must exist");
+  }
+
+  // In the following code block, we extract the values of `__cuda_array_interface__`
+  std::vector<int64_t> sizes;
+  ScalarType dtype;
+  npy_intp data_ptr;
+  std::vector<int64_t> strides;
+  try {
+    if (!PyDict_Check(cuda_dict)) {
+      throw TypeError("`__cuda_array_interface__` must be a dict");
+    }
+    PyObject *py_shape = PyDict_GetItemString(cuda_dict, "shape");
+    PyObject *py_typestr = PyDict_GetItemString(cuda_dict, "typestr");
+    PyObject *py_data = PyDict_GetItemString(cuda_dict, "data");
+    if (py_shape == nullptr || py_typestr == nullptr || py_data == nullptr) {
+      throw TypeError("attributes `shape`, `typestr`, and `data` must exist");
+    }
+
+    // Extract the `shape` attribute
+    int ndim = PySequence_Length(py_shape);
+    if (ndim == -1) {
+      throw TypeError("shape must be a sequence");
+    }
+    std::vector<npy_intp> npy_shape(ndim);
+    PyArray_IntpFromSequence(py_shape, npy_shape.data(), ndim);
+    sizes = to_aten_shape(ndim, npy_shape.data());
```

Comment:
That would save you from having to do another conversion to actually get into ATen sizes.